cdap-watchdog/src/main/java/co/cask/cdap/logging/framework/distributed/DistributedLogFramework.java

```diff
@@ -0,0 +1,179 @@
+        checkpointManagerFactory.create(topic, pipelineSpec.getCheckpointPrefix()), brokerService, config));
+    }
+
+    // Returns a Service that start/stop all pipelines.
+    return new AbstractIdleService() {
+      @Override
+      protected void startUp() throws Exception {
+        // Starts all pipeline
+        validateAllFutures(Iterables.transform(pipelines, new Function<Service, ListenableFuture<State>>() {
+          @Override
+          public ListenableFuture<State> apply(Service service) {
+            return service.start();
+          }
+        }));
+      }
+
+      @Override
+      protected void shutDown() throws Exception {
+        // Stops all pipeline
+        validateAllFutures(Iterables.transform(pipelines, new Function<Service, ListenableFuture<State>>() {
+          @Override
+          public ListenableFuture<State> apply(Service service) {
+            return service.stop();
+          }
+        }));
+      }
+    };
+  }
+
+  /**
+   * Blocks and validates all the given futures completed successfully.
+   */
+  private void validateAllFutures(Iterable<? extends ListenableFuture<?>> futures) throws Exception {
+    // The get call shouldn't throw exception. It just block until all futures completed.
+    Futures.successfulAsList(futures).get();
+
+    // Iterates all futures to make sure all of them completed successfully
+    Throwable exception = null;
+    for (ListenableFuture<?> future : futures) {
+      try {
+        future.get();
+      } catch (ExecutionException e) {
+        if (exception == null) {
+          exception = e.getCause();
+        } else {
+          exception.addSuppressed(e.getCause());
+        }
+      }
+    }
+
+    // Throw exception if any of the future failed.
+    if (exception != null) {
+      if (exception instanceof Exception) {
+        throw (Exception) exception;
+      }
+      throw new RuntimeException(exception);
+    }
+  }
+
+  /**
+   * Determines the buffer size for one pipeline.
+   */
+  private long getBufferSize(int numberOfPipelines, CConfiguration cConf, int partitions) {
+    long bufferSize = cConf.getLong(Constants.Logging.PIPELINE_BUFFER_SIZE);
+    if (bufferSize > 0) {
+      return bufferSize;
+    }
+
+    double bufferRatio = cConf.getDouble(Constants.Logging.PIPELINE_AUTO_BUFFER_RATIO);
+    Preconditions.checkArgument(bufferRatio > 0 && bufferRatio < 1,
+                                "Config %s must be between 0 and 1", Constants.Logging.PIPELINE_AUTO_BUFFER_RATIO);
+
+    int kafkaFetchSize = cConf.getInt(Constants.Logging.PIPELINE_KAFKA_FETCH_SIZE) * partitions;
+
+    // Try to derive it from the total memory size, the number of pipelines and the kafka fetch size
+    bufferSize = (long) ((Runtime.getRuntime().maxMemory() * bufferRatio - kafkaFetchSize) / numberOfPipelines);
+    // The size has to be > 0 for it to make any progress. This is just to safe guard
+    return bufferSize > 0 ? bufferSize : 1L;
```

Comment:
if buffer size is 1, event queue size would be 1 and we wouldn't be buffering, but it would still work. so its okay.